lib/types.js

```diff
@@ -214,18 +224,43 @@ function generateTypeConversion(
       `);
     }
 
-    // Do not convert buffer source types as the impl code can either "get a reference" or "get a copy" to the bytes.
     if (union.ArrayBuffer || union.object) {
-      output.push(`if (utils.isArrayBuffer(${name})) {}`);
+      if (union.ArrayBuffer) {
+        output.push(`if (utils.isArrayBuffer(${name})) {
+          ${generateTypeConversion(ctx, name, union.ArrayBuffer, [], parentName, errPrefix).body}
+        }`);
+      } else if (union.object) {
+        output.push(`if (utils.isArrayBuffer(${name})) {}`);
+      }
+    }
+    if (union.SharedArrayBuffer || union.object) {
+      if (union.SharedArrayBuffer) {
+        output.push(`if (utils.isSharedArrayBuffer(${name})) {
+          ${generateTypeConversion(ctx, name, union.SharedArrayBuffer, [], parentName, errPrefix).body}
+        }`);
+      } else if (union.object) {
+        output.push(`if (utils.isSharedArrayBuffer(${name})) {}`);
+      }
     }
     if (union.ArrayBufferViews.size > 0 || union.object) {
-      let condition = `ArrayBuffer.isView(${name})`;
-      // Skip specific type check if all ArrayBufferView member types are allowed.
-      if (union.ArrayBufferViews.size !== arrayBufferViewTypes.size) {
-        const exprs = [...union.ArrayBufferViews].map(a => `${name}.constructor.name === "${a}"`);
-        condition += ` && (${exprs.join(" || ")})`;
+      if (union.ArrayBufferViews.size > 0) {
+        const viewIdlTypes = new Set([...union.ArrayBufferViews].map(item => item.idlType));
+        // Skip specific type check if all ArrayBufferView member types are allowed.
+        if (viewIdlTypes.size === arrayBufferViewTypes.size) {
+          const viewExtAttrs = mergeExtAttrsOfTypes(union.ArrayBufferViews);
+          output.push(`if (ArrayBuffer.isView(${name})) {
+            ${generateWebIDLConversions(`conversions["ArrayBufferView"]`, viewExtAttrs)}
```

Comment:
This bit is slightly awkward. I can't use `generateTypeConversion` with IDL type `ArrayBufferView`, since that will just resolve the union again and recurse back to this same line. So I have to jump straight into `generateWebIDLConversions` instead.